teslajsonpy/homeassistant/homelink.py

```diff
@@ -0,0 +1,90 @@
+#  SPDX-License-Identifier: Apache-2.0
+"""
+Python Package for controlling Tesla API.
+
+For more details about this api, please refer to the documentation at
+https://github.com/zabuldon/teslajsonpy
+"""
+
+from teslajsonpy.exceptions import HomelinkError
+from teslajsonpy.homeassistant.vehicle import VehicleDevice
+
+
+class TriggerHomelink(VehicleDevice):
+    """Home-Assistant class for trigger homelink of Tesla vehicles."""
+
+    def __init__(self, data, controller):
+        """Initialize the trigger homelink for the vehicle.
+
+        Parameters
+        ----------
+        data : dict
+            The trigger homelink for a Tesla vehicle.
+            https://tesla-api.timdorr.com/vehicle/commands/homelink
+        controller : teslajsonpy.Controller
+            The controller that controls updates to the Tesla API.
+
+        Returns
+        -------
+        None
+
+        """
+        super().__init__(data, controller)
+        self.type = "trigger homelink"
+        self.hass_type = "button"
+        self.name = self._name()
+        self.uniq_name = self._uniq_name()
+        self.__longitude = None
+        self.__latitude = None
+        self.__homelink_device_count = None
+        self.__homelink_nearby = None
+        self.__homelink_available = False
+
+    async def async_update(self, wake_if_asleep=False, force=False):
+        """Update the trigger homelink of the vehicle."""
+        await super().async_update(wake_if_asleep=wake_if_asleep)
+        self.refresh()
+
+    @staticmethod
+    def has_battery() -> bool:
+        """Return whether the device has a battery."""
+        return False
+
+    async def trigger_homelink(self) -> None:
+        """Trigger Homelink."""
+        if self.__latitude is not None and self.__longitude is not None:
+            if not self.__homelink_device_count:
+                raise HomelinkError(f"No homelink devices added to {self.car_name()}.")
+            if not self.__homelink_nearby:
+                raise HomelinkError(f"No homelink devices near {self.car_name()}.")
+            await self._controller.api(
```

Comment:
Are there any other homelink specific errors we should consider? For example, what if the lat/long is not near the home? I haven't played with the homelink service so don't know what errors it could produce.

Reply:
From what I have experienced in my own testing, I've only been able to get a `too_far_from_vehicle` error from the api. I'm sure there's a "no homelink devices added" response but I can't easily test that, I'd have to remove and relearn all of my homelink devices.

Rather than relying on a failed api response error, my approach was to preemptively raise these errors.
Also, since we're pulling lat/lon straight from the car, as long as the car reports a device is nearby, we can be sure we're sending a valid lat/lon.

However I'm not sure if async_update (and thus refresh) run automatically when the button is pressed. If not, there is a possible edge case:
- the latest update reports device nearby and has valid lat/lon; but car has since moved and no longer near a homelink device
  - this would cause the api call to get a response with `{'response': {'reason': 'too_far_from_vehicle', 'result': False}}`, which I'm not sure if the controller would notice this as an error and bubble it up or not. I'd have to manually fudge the code to test this on my car and see what happens. I was under the impression it would get noticed by the controller that the api call returned an error, but now I'm not sure.

Reply:
> Rather than relying on a failed api response error, my approach was to preemptively raise these errors.

While it's ok to do this, we should also handle the api errors too even if wrapped up generically. Perhaps see if the result is False and then pass the reason if it exist.

Reply:
I added a response check and raise a HomelinkError in case the response indicates the action wasn't performed to bubble up the error (no matter what it is) to HA logs.

Reply:
While not necessary since we're at teslajsonpy, I haven't checked, but I assume you are generating a HA log event in the HA component.

Reply:
correct, in HA it's being caught and logged as error.